### ISSUE TITLE
Bin nomenclature not consistent

### ISSUE THREAD

**Issue by fpusan** [jtamames/SqueezeMeta#140]
Hi Fernando, 

I just stumbled across an issue after using the subsetORFs function, maybe it is related to the error that you reported here.

I am trying to create a combined table containing the ORF, contig and bin information for some ORFs of interest, but there is an issue with the Bin IDs.

Here is what I do:
```

# make a vector containing the orf names
test_orfs <- c("megahit_363190_1-96", "megahit_413329_895-1317", "megahit_458224_2-187", "megahit_605321_2-169", "megahit_733519_1958-2851", "megahit_966690_1138-1290", "megahit_1978256_3-206", "megahit_2338645_3-491", "megahit_2342939_23546-23749", "megahit_2503433_63512-64405", "megahit_2512696_20560-20796")
# make subset with test_orfs
test_subset = subsetORFs(so245, orfs = test_orfs)

test_bin_table = test_subset$bins$table
row.names(test_bin_table)
```
[1] "concoct.28.fa_sub.contigs"     "maxbin.0783.fasta_sub.contigs" "maxbin.0345.fasta_sub.contigs" "metabat2.1630.fa_sub.contigs" 
[5] "metabat2.430.fa.contigs"

`test_subset$contigs$bins`
               DASTool                        
megahit_363190  "No_bin"                       
megahit_413329  "No_bin"                       
megahit_458224  "No_bin"                       
megahit_605321  "No_bin"                       
megahit_733519  "concoct.28.fa_sub.contigs"    
megahit_966690  "No_bin"                       
megahit_1978256 "maxbin.0783.fasta_sub.contigs"
megahit_2338645 "maxbin.0345.fasta_sub.contigs"
megahit_2342939 "metabat2.1630.fa_sub.contigs" 
megahit_2503433 "maxbin.0345.fasta_sub.contigs"
megahit_2512696 "metabat2.430.fa.contigs"  

`test_subset$contigs$table$`Bin ID``
[1] ""                              ""                              ""                              ""                             
 [5] "concoct.28.fa_sub.contigs"     ""                              "maxbin.0783.fasta_sub.contigs" "maxbin.0345.fasta_sub.contigs"
 [9] "metabat2.1630.fa_sub.contigs"  "maxbin.0345.fasta_sub.contigs" "metabat2.430" 

Out of some reason, the name of one bin is different in `test_subset$contigs$table$`Bin ID`` compared to `test_subset$contigs$bins` and `test_subset$bins$table`, metabat2.430 instead of the correct metabat2.430.fa.contigs.

This becomes a problem when I try to create a combined table:

```
# now merge the contigs table with the bin table
test_subset_contig_bin_table <- merge.data.frame(test_subset$contigs$table, test_subset$bins$table, all.x = TRUE, by.x="Bin ID", by.y = 'row.names')
subset_table <- subset(test_subset_contig_bin_table, select = c('Bin ID', 'Tax.y', 'Num contigs', 'Completeness'))
subset_table
```
                       
 ```
 Bin ID                                                                                  Tax.y Num contigs Completeness
1                                                                                                                  <NA>          NA           NA
2                                                                                                                  <NA>          NA           NA
3                                                                                                                  <NA>          NA           NA
4                                                                                                                  <NA>          NA           NA
5                                                                                                                  <NA>          NA           NA
6      concoct.28.fa_sub.contigs                                                                             k_Bacteria        1502        35.53
7  maxbin.0345.fasta_sub.contigs         k_Bacteria;p_Proteobacteria;n_delta/epsilon subdivisions;c_Deltaproteobacteria         915        57.73
8  maxbin.0345.fasta_sub.contigs         k_Bacteria;p_Proteobacteria;n_delta/epsilon subdivisions;c_Deltaproteobacteria         915        57.73
9  maxbin.0783.fasta_sub.contigs       k_Bacteria;n_PVC group;p_Verrucomicrobia;c_Verrucomicrobiae;o_Verrucomicrobiales        1562        54.34
10  metabat2.1630.fa_sub.contigs k_Bacteria;p_Proteobacteria;c_Alphaproteobacteria;o_Rhodobacterales;f_Rhodobacteraceae         154        35.85
11                  metabat2.430                                                                                   <NA>          NA           NA

```
Obviously, because the bin IDs are different in the two tables I want to merge, there is no bin information in the merged table for bin metabat2.430.fa.contigs.

Do you have an idea about what goes wrong here and how I can potentially fix it?

Thanks a lot

_Originally posted by @bresyd in https://github.com/jtamames/SqueezeMeta/issues/138#issuecomment-673521703_

**Comment by fpusan** [comment]
@bresyd, just moved the discussion to its own issue.
Could you please send me the `/path/to/your/project/results/19.*.bintable` file?

**Comment by bresyd** [comment]
Hi Fernando,

thanks for getting back to me on that issue. I just sent you the bintable file on your gmail email.

Cheers

benni

**Comment by fpusan** [comment]
The nomenclature of the _metabat2.430_ bin is correct in the bin table produced by SqueezeMeta, so the problem indeed appears to be my fault. Sigh.

Is the name of the bin also wrong right after you load the data into SQMtools, or does it break only after performing a subset?

**Comment by bresyd** [comment]
I think the issue is already present in the object right after loading it into SQM tools:


```
so245 = loadSQM('/home/btschits/SO245_metagenomes/squeezemeta_min1000/r_directories/maxbin-metabat-concoct/sqm_so245_metagenome_min1000', engine = 'data.table')

metabat2.430.fa.contigs_contigs = rownames(so245$contigs$table) [so245$contigs$table$'Bin ID'=='metabat2.430.fa.contigs']
metabat2.430.fa.contigs_contigs
```
character(0)

```
metabat2.430_contigs = rownames(so245$contigs$table) [so245$contigs$table$'Bin ID'=='metabat2.430']
metabat2.430_contigs
```
 [1] "megahit_84213"   "megahit_111457"  "megahit_235250"  "megahit_449451"  "megahit_464911"  "megahit_468760"  "megahit_515706"  "megahit_527729"  "megahit_665743" 
[10] "megahit_770102"  "megahit_873486"  "megahit_971061"  "megahit_1090999" "megahit_1113264" "megahit_1139011" "megahit_1240532" "megahit_1254269" "megahit_1293296"
[19] "megahit_1323391" "megahit_1483070" "megahit_1550197" "megahit_1711936" "megahit_1747619" "megahit_1890768" "megahit_2013466" "megahit_2116520" "megahit_2116842"
[28] "megahit_2307224" "megahit_2312535" "megahit_2314547" "megahit_2316199" "megahit_2318313" "megahit_2323736" "megahit_2333008" "megahit_2337736" "megahit_2339046"
[37] "megahit_2344588" "megahit_2349821" "megahit_2368363" "megahit_2368377" "megahit_2372458" "megahit_2383283" "megahit_2399310" "megahit_2426537" "megahit_2431934"
[46] "megahit_2433225" "megahit_2443147" "megahit_2455175" "megahit_2456606" "megahit_2472759" "megahit_2477576" "megahit_2478125" "megahit_2488306" "megahit_2506696"
[55] "megahit_2509264" "megahit_2512696" "megahit_2515870" "megahit_2515875" "megahit_2537040" "megahit_2543032" "megahit_2566448" "megahit_2570307" "megahit_2575297"
[64] "megahit_2579554"

**Comment by fpusan** [comment]
Can you also send me the `/path/to/your/project/results/19.*.contigtable` file?

**Comment by fpusan** [comment]
Sorry, the right one would be `/path/to/your/project/results/20.*.contigtable`.
Please also the `/path/to/your/project/intermediate/19.*.contigsinbins` file, which I suspect is to blame.

**Comment by bresyd** [comment]
Hi, 

here is the contigsinbin file
[19.sqm_so245_metagenome_min1000.contigsinbins.gz](https://github.com/jtamames/SqueezeMeta/files/5090602/19.sqm_so245_metagenome_min1000.contigsinbins.gz)

the other file has 139Mb (zipped); any idea how I can transfer that to you? Or would a subset of the table be sufficient?

**Comment by fpusan** [comment]
Just grep the offending bin

`grep metabat2.430 20.*.contigtable`

**Comment by bresyd** [comment]
Hi,

here is the subset of the contigtable using `grep metabat2.430 20.*.contigtable > metabat2.430_subset.contigtable`

[metabat2.430_subset.contigtable.gz](https://github.com/jtamames/SqueezeMeta/files/5094667/metabat2.430_subset.contigtable.gz)

**Comment by fpusan** [comment]
The bin name is indeed wrong in the contig table as generated by the `20.getcontigs.pl` script from SqueezeMeta. We'll have to fix that.

However, the `contigsinbins` file has the right information, which is stored in `SQM$contigs$bins` when loading the project into R with SQMtools.

From your original post I understand that you want to obtain a table containing the ORF, contig and bin information for some ORFs of interest (this must be right bc I just copy-pasted it).

Would this suffice?

```
# Let SQM be a SQM object containing your project.
# Let orfnames be a character vector with the orfs of interest.
orfs = SQM$orfs$table[orfnames,]
contignames = SQM$orfs$table[orfnames, 'Contig ID']
contigs = SQM$contigs$table[contignames,]
binnames = SQM$contigs$bins[contignames,]
bins = SQM$bins$table[binnames,]
merged = cbind(orfs, contigs, bins)
```

You'll possibly find a fancier way to do it, and might want to rename some column names, but the info is there. The key point is to use `SQM$contigs$bins` instead of `SQM$contigs$table[,'Bin ID']` until we get that bug fixed.

Let me know if this works for you!

**Comment by fpusan** [comment]
Closing due to lack of activity
